transport/src/main/java/io/netty/channel/nio/NioEventLoop.java

```diff
@@ -372,6 +373,21 @@ public int registeredChannels() {
         return selector.keys().size() - cancelledKeys;
     }
 
+    @Override
+    public Iterator<AbstractNioChannel> registeredChannelsIterator() {
+        List<AbstractNioChannel> nioChannelList = new ArrayList<AbstractNioChannel>();
```

Comment:
The goal of original issue is to avoid wasted memory/redundant data structures, so here we'd need to iterate directly over SelectionKey collection, and that iterator needs to be read-only.